[Timothy-Ho0203] iP

Timothy-Ho0203 opened the pull request:
# TommyTalks
> "Your mind is for having ideas, not storing them."  - David Allen [(source)](https://dansilvestre.com/productivity-quotes/)

TommyTalks frees your mind of having to remember all these ideas, it'll do it for you! It's,
- Text-based
- Easy to learn
- ~~FAST~~ ***SUPER FAST*** to use

All you need to do is, 
1.  Download `TommyTalks.jar` from [here](https://github.com/Timothy-Ho0203/ip/releases)
2.  Run `java -jar TommyTalks.jar` in your command prompt
3.  Enjoy 😄 

Features:
- [x] Managing tasks, deadlines and events
- [ ] Reminders (Coming soon)
- [ ] Timeline recommendations (Coming soon)
---
## Program
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public class TommyTalks {
    public static void main(String[] args) {
        App chatBot = new App();
        chatBot.run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Timothy-Ho0203 committed:
refactor: add new class to store helper functions

Timothy-Ho0203 committed:
feat: add feature for bot to echo instructions

Timothy-Ho0203 committed:
feat: Add feature to store text and display back to user when requested

Timothy-Ho0203 committed:
feat: Add ability to mark task as done or undone

refactor: Add new class Task to organise tasks

Timothy-Ho0203 committed:
refactor: Added 3 classes to extend Task class for specific behaviour

Timothy-Ho0203 committed:
feat: Add Automatic Text UI Testing feature

Timothy-Ho0203 committed:
feat: Add error handling and messages

Timothy-Ho0203 committed:
feat: Add support for deleting tasks from list

Timothy-Ho0203 committed:
refactor: Add packages for Task and Exceptions, Optimize error handling, Add help feature

Timothy-Ho0203 committed:
Add Storage.java

TommyTalks does not save tasks, causing users to restart their lists of
tasks every session.

Adding a storage class to keep track of tasks improves user-experience
since it makes the app more practical.

Timothy-Ho0203 committed:
Merge branch 'branch-Level-7'
Add Storage.java class

Changes have been made in branch-Level-7 regarding data storage.

To update master branch, we will do a merge commit with branch-Level-7.

Timothy-Ho0203 committed:
Update Tasks to take in date and time objects

Tasks track dates/times as Strings.

Let's update the app to track dates/times better so that it can track
users' tasks more accurately.

Timothy-Ho0203 committed:
Merge branch 'branch-Level-8'
Update Tasks to be able to parse date/time

Changes have been made in branch-Level-8 regarding the storage and
understanding of date/time in the app.

To update master branch, we will do a merge commit with branch-Level-8.

Timothy-Ho0203 committed:
Refactor App for more OOP

There is repeated and similar code when determining input commands.

Let's create a seperate Command class handle the actions outside of App
class to imrpove readability of code.

Timothy-Ho0203 committed:
Merge branch 'branch-A-MoreOOP'
Refactor code in App class

Changes have been made in branch-A-MoreOOP to create new Command classes
to improve readability of the code.

To update master branch, we will do a merge commit with
branch-A-MoreOOP.

Timothy-Ho0203 committed:
Organise classes into packages

Timothy-Ho0203 committed:
Merge branch 'branch-A-Packages'
Organise classes into packages

Timothy-Ho0203 committed:
Merge branch 'add-gradle-support'
Automate build tasks of the project

Timothy-Ho0203 committed:
Add Gradle support to the project

Timothy-Ho0203 committed:
Merge branch 'branch-A-Gradle'
Add Gradle support to the project

Timothy-Ho0203 committed:
Add JUnit tests

Timothy-Ho0203 committed:
Merge branch 'branch-A-JUnit'
Add JUnit tests

Timothy-Ho0203 committed:
Create JAR file

Timothy-Ho0203 committed:
Merge branch 'A-Jar'
Create JAR file

Timothy-Ho0203 committed:
Add find feature for the app

Users cannot find tasks by words and it is impractical for them to
remember the exact index of the task stored

Let's add a find command to allow users to search for tasks by words

Timothy-Ho0203 committed:
Add JavaDocs

Timothy-Ho0203 committed:
Refactor code to follow coding standards

Timothy-Ho0203 committed:
Refactor directory that data is saved to

Timothy-Ho0203 committed:
Refactor directory that data is saved to

Timothy-Ho0203 committed:
Merge branch 'branch-Level-9'
Add find feature

Find feature has been added in branch-Level-9.

To update master branch, we will do a merge commit with branch-Level-9.

Timothy-Ho0203 committed:
Resolved merge conflicts from branch-A-JavaDoc

Timothy-Ho0203 committed:
Resolve merge conflicts from branch-A-CodingStandard

LonelyFort commented on src/main/java/TommyTalks/Ui.java:
```diff
@@ -0,0 +1,122 @@
+     * Prints the error message specified by e.
+     * @param e Error that was caught.
+     */
+    public void errorMessage(Exception e) {
```
could refactor to "printErrorMessage" to fit the code quality

LonelyFort commented on text-ui-test/input.txt:
put this under .gitignore

LonelyFort commented on text-ui-test/EXPECTED.TXT:
put this under .gitignore

LonelyFort commented on src/main/java/TommyTalks/Ui.java:
```diff
@@ -0,0 +1,122 @@
+        System.out.print(LINE);
+    }
+
+    public void setStupid() {
+        isStupid = true;
+    }
+
+    /**
+     * Prints message to be displayed when user inputs invalid commands.
+     * @throws InvalidFormatException If input is of wrong format.
+     */
+    public void invalidCommand() {
+        if (!isStupid) {
+            String err = """
+                    \to.O What's that? Please specify
+                    \tthe type of event(todo, deadline,
+                    \tevent etc) in the appropriate
+                    \tformats. I would appreciate it!""";
+            throw new InvalidFormatException(err);
+        } else {
+            String help = """
+                    \tYou seem a lil lost, would you
+                    \tlike to view all possible
+                    \tcommands using help?""";
+            throw new InvalidFormatException(help);
+        }
+    }
```
too real 😂

LonelyFort commented on src/main/java/TommyTalks/Storage.java:
```diff
@@ -0,0 +1,215 @@
+            }
+        }
+        // Populate tasks
+        try (BufferedReader reader = new BufferedReader(new FileReader(f))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                addToList(line);
+            }
+        } catch (IOException e) {
+            System.out.println("Error reading file: " + e.getMessage());
+        }
+    }
```
look I'm not here for code standard but I'm not competent enough to know what does the parameter in the try block do so, maybe don't do it like this?

wenruu replied:
seems pretty readable, just adding lines in the file to a list until theres no more next line?

LonelyFort commented on src/main/java/TommyTalks/Storage.java:
```diff
@@ -0,0 +1,215 @@
+     */
+    public void saveToFile(Task task) {
+        tasks.add(task);
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filePath, true))) {
+            writer.append(task.getKeyInfo());
+            writer.newLine();
+        } catch (IOException e) {
+            System.out.println("Error writing to file: " + e.getMessage());
+        }
+        String base = """
+            Added:
+            %s
+            You have %d tasks in the list.
+        """;
+        int size = tasks.size();
+        String result = String.format(base, task, size);
+        System.out.print(result + LINE);
+    }
```
same issue as above.

LonelyFort commented on src/main/java/TommyTalks/Storage.java:
```diff
@@ -0,0 +1,215 @@
+    /**
+     * Saves the list of tasks in file to hard disk.
+     */
+    public void save() {
```
wait shouldn't this be named saveToFile instead if you are saving INTO a text file for this one

Timothy-Ho0203 replied:
the saveToFile was for writing to the text file, although save is also writing to text file, it's sort of the final save before closing the program hence a more generic "save" naming

LonelyFort reviewed (commented):
Code quality review: 
Overall no big issue naming-wise. "Try" not to put a parameter for a try block in Storage.java file, otherwise well done!

Timothy-Ho0203 committed:
Add GUI for TommyTalks

Timothy-Ho0203 committed:
Add checkstyle for gradle

Timothy-Ho0203 committed:
Add assertions

There are no assertions in the source code.

Adding insertions can help to reduce bugs by ensuring assumptions made
by programmer is held throughout the execution of the program.

Let's add some assertions at key points throughout the code to document
these assumptions.

Using assertions is easy to integrate into our code to ensure our
assumptions hold true.

Assertions have been added in Storage and TaskCommand class.

Timothy-Ho0203 committed:
Refactor code to fit Code Quality standards

Code does not follow Code Quality standards closely.

Code that is written to different standards can make it harder to read
and follow along when being reviewed by peers, hence we should aim to
improve the readability of the code and follow an agreed-upon coding
standard.

Let's refactor the code to follow the SE-EDU Java coding standard.

The SE-EDU Java coding standard is the agreed standard to follow by the
course, so we should follow this instead of relying on our intuition.

Refer to the SE-EDU Java coding standards at
https://se-education.org/guides/conventions/java/intermediate.html

Timothy-Ho0203 committed:
Merge pull request #2 from Timothy-Ho0203/branch-A-CodeQuality

Refactor code to fit Code Quality standards

Timothy-Ho0203 committed:
Merge branch 'master' of https://github.com/Timothy-Ho0203/ip into branch-A-Assertions
Merge master to branch-A-Assertions

Branch-A-Assertions is out of date with master branch after master
merged with branch-A-CodeQuality.

To continue working on branch-A-Assertion, we should update
branch-A-Assertions to reduce conflicts.

Let's merge master to branch-A-Assertion.

This is to update branch-A-Assertion in the cleanest way possible.

Timothy-Ho0203 committed:
Add more assertions

There are not much assertions present.

We may not be addressing the implicit assumptions enough, so we should
include more checks throughout the code.

Let's add some more assertions throughout the code.

Assertions provide a better way to check for underlying conditions
compared to if else statements.

Timothy-Ho0203 committed:
Merge pull request #1 from Timothy-Ho0203/branch-A-Assertions

Add assertions

Timothy-Ho0203 committed:
Add ability to prioritise tasks

Tasks were ordered based on order of insertion, and could not be edited
further.

User may have preferences on the ordering of the tasks, thus they may
want to be able to order the tasks according to their preferences.

Let's add a priority attribute to Task objects, and allow for users to
be able to sort the current task list based on this attribute.

Using the comparator interface allows for tasks objects to have a
natural ordering, allowing for easier sorting.

On top of adding sort and priority feature, also modified .gitignore to
stop tracking test txt files.

Timothy-Ho0203 committed:
Add Ui.png to docs

Timothy-Ho0203 committed:
Add user guide

Timothy-Ho0203 committed:
Refactor Commands to handle specific commands directly

Timothy-Ho0203 committed:
Fix Main-Class entry

Timothy-Ho0203 committed:
Update README.md

Timothy-Ho0203 committed:
Add more JUnit tests
